File: src/SeedSource/SeedSource.js
// JavaScript implementation of SeedSource.GetSeed (bv64 -> BigInt).
// Build/run:  dafny run --target:js SeedSourceDemo.dfy --input SeedSource.js
// (Dafny's JS runtime needs bignumber.js available, e.g. `npm install bignumber.js`.)
//
// The generated module does `let $module = SeedSource;`, i.e. it expects a
// pre-existing global `SeedSource`; we define it here with the GetSeed member.
let SeedSource = (function () {
  let $module = {};
  $module.GetSeed = function () {
    const crypto = require('crypto');
    return crypto.randomBytes(8).readBigUInt64LE(0);   // bv64 is a BigInt in Dafny's JS runtime
  };
  return $module;
})();

File: src/SeedSource/SeedSource.go
// Go implementation of SeedSource.GetSeed (bv64 -> uint64).
// Build/run:  dafny run --target:go SeedSourceDemo.dfy --input SeedSource.go
// (Requires `goimports` on PATH — `go install golang.org/x/tools/cmd/goimports@latest`.)
package SeedSource

import (
	"crypto/rand"
	"encoding/binary"
	"time"
)

func GetSeed() uint64 {
	var b [8]byte
	_, _ = rand.Read(b[:])
	return binary.LittleEndian.Uint64(b[:])
}

// NowNanos returns a nanosecond timestamp; only differences are consumed, so
// the origin (Unix epoch here) is unspecified as far as callers are concerned.
func NowNanos() uint64 {
	return uint64(time.Now().UnixNano())
}
